lib/AppURLHider.js

```diff
@@ -0,0 +1,104 @@
+/* globals Services */
+"use strict";
+
+const {Cu, Ci} = require("chrome");
+const tabs = require("sdk/tabs");
+
+Cu.import("resource://gre/modules/Services.jsm");
+
+function AppURLHider(appURLs) {
+  this._appURLs = appURLs;
+
+  this.maybeHideURL = this.maybeHideURL.bind(this);
+  tabs.on("open", this.maybeHideURL);
+  tabs.on("ready", this.maybeHideURL);
+  tabs.on("load", this.maybeHideURL);
+  tabs.on("activate", this.maybeHideURL);
+  tabs.on("pageshow", this.maybeHideURL);
```

Comment:
It might just be `activate` because the rest should be handled by `gInitialPages` and the route change handler. I'll play with it.